File: src/main/java/io/gittul/infra/auth/jwt/JwtProvider.java
package io.gittul.infra.auth.jwt;

import io.gittul.infra.auth.exception.AuthenticationException;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Date;

@Component
public class JwtProvider {

    @Value("${jwt.secret}")
    private String secretKey;

    @Value("${jwt.expiration}")
    private long expirationTime;


    /**
     * JWT 토큰 생성
     */
    public String createToken(TokenUserInfo userInfo) {
        Date now = new Date();
        Date accessExpiration = new Date(now.getTime() + this.expirationTime);

        return Jwts.builder()
                .subject(userInfo.userName())
                .claim("userId", userInfo.userId())
                .claim("email", userInfo.email())
                .issuedAt(now)
                .expiration(accessExpiration)
                .signWith(getKey(), Jwts.SIG.HS256)
                .compact();
    }

    public TokenUserInfo validateToken(String token) {
        try {
            return parseToken(token);
        } catch (Exception e) {
            if (e instanceof ExpiredJwtException) {
                throw AuthenticationException.EXPIRED_TOKEN;
            }

            throw AuthenticationException.INVALID_TOKEN;
        }
    }

    private TokenUserInfo parseToken(String token) {
        Claims claims = Jwts.parser()
                .verifyWith(getKey())
                .build()
                .parseSignedClaims(token)
                .getPayload();

        System.out.println(claims);

        return new TokenUserInfo(
                claims.get("userId", Long.class),
                claims.getSubject(), // userName
                claims.get("email", String.class)
        );
    }

    private SecretKey getKey() {
        return Keys.hmacShaKeyFor(secretKey.getBytes());
    }
}

File: app/src/main/kotlin/io/gittul/app/infra/auth/jwt/JwtProvider.kt
package io.gittul.app.infra.auth.jwt

import com.fasterxml.jackson.databind.ObjectMapper
import io.gittul.app.infra.auth.exception.AuthenticationException
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtProvider(
    private val objectMapper: ObjectMapper = ObjectMapper()
) {
    @Value("\${jwt.secret}")
    private val secretKey: String? = null

    @Value("\${jwt.expiration}")
    private val expirationTime: Long = 0


    /**
     * JWT 토큰 생성
     */
    fun createToken(userInfo: TokenUserInfo): String {
        val now = Date()
        val accessExpiration = Date(now.time + this.expirationTime)

        val userInfoJson: String = objectMapper.writeValueAsString(userInfo)

        return Jwts.builder()
            .claim("userInfo", userInfoJson)
            .issuedAt(now)
            .expiration(accessExpiration)
            .signWith(this.key, Jwts.SIG.HS256)
            .compact()
    }

    fun validateToken(token: String): TokenUserInfo {
        try {
            return parseToken(token)
        } catch (e: Exception) {
            println(e)
            if (e is ExpiredJwtException) {
                throw AuthenticationException.EXPIRED_TOKEN
            }
            throw AuthenticationException.INVALID_TOKEN
        }
    }

    private fun parseToken(token: String): TokenUserInfo {
        val claims = Jwts.parser()
            .verifyWith(this.key)
            .build()
            .parseSignedClaims(token)
            .payload

        val userInfoJson: String = claims.get("userInfo", String::class.java)
        return objectMapper.readValue(userInfoJson, TokenUserInfo::class.java)
    }

    private val key: SecretKey
        get() = Keys.hmacShaKeyFor(secretKey!!.toByteArray())
}
